synapse/storage/end_to_end_keys.py

```diff
@@ -85,11 +89,12 @@ def _get_e2e_device_keys_txn(
             "    k.key_json"
             " FROM devices d"
             "    %s JOIN e2e_device_keys_json k USING (user_id, device_id)"
-            " WHERE %s"
+            " WHERE (%s) AND NOT COALESCE(d.hidden, ?)"
```

Comment:
having looked at how this works out, I'm really sorry but I think I'm going to revise my original comments here; it's just too awkward. I think the 30s or so it's going to take to add a NOT NULL column to the table will be worth it for not jumping through these hoops and having to manage a tristate.

Really sorry for vacillating on this.

@erikjohnston says:

>  I'd vote either a) make the column NOT NULL or b) make the column something like an enum and so you have NULL or "hidden" or something (which really only makes sense if there is potential for that column to be extensible)